File: java/10CompleteImplementationCode/src/main/java/com/neomatrix369/services/WordsFrequencyReporter.java
package com.neomatrix369.services;

import com.neomatrix369.domain.WordEntity;
import com.neomatrix369.sorter.SortByFrequencyDescAndByWordAlphabetically;

import java.util.Collections;
import java.util.List;

public class WordsFrequencyReporter {
    private static final String EACH_LINE_REPORT_FORMAT = "%d %s" + "\n";

    public StringBuilder parse(int topNFrequency, String wordsList) {
        StringBuilder report = new StringBuilder();

        if ( checkIfWordsListIsNotAssignedOrEmpty(wordsList) ) {
            return report;
        }

        WordsFrequencyList wordsFrequencyHelper = new WordsFrequencyList();
        List<WordEntity> wordsFrequencyList =
                wordsFrequencyHelper.calculateWordsFrequency(wordsList);

        Collections.sort(
                wordsFrequencyList, new SortByFrequencyDescAndByWordAlphabetically());

        return generateReportToReturn(report, wordsFrequencyList, topNFrequency);
    }

    private boolean wordsListIsNotAssigned(String wordsList) {
        return wordsList == null;
    }

    private StringBuilder generateReportToReturn(
            StringBuilder report,
            List<WordEntity> wordsFrequencyList,
            int topNFrequency) {
        if (topNFrequency <= 0) {
            return report;
        }

        topNFrequency = keepTopNFrequencyWithinRange(wordsFrequencyList, topNFrequency);

        List<WordEntity> filteredList = wordsFrequencyList.subList(0, topNFrequency);
        for(WordEntity wordEntity: filteredList) {
            report.append(String.format(EACH_LINE_REPORT_FORMAT,
                    wordEntity.getFrequency(),
                    wordEntity.getWord()));
        }

        return report;
    }

    private int keepTopNFrequencyWithinRange(List<WordEntity> wordsFrequencyList, int topNFrequency) {
        if (topNFrequency > wordsFrequencyList.size()) {
            topNFrequency = wordsFrequencyList.size();
        }
        return topNFrequency;
    }

    private boolean checkIfWordsListIsNotAssignedOrEmpty(String wordsList) {
        return wordsListIsNotAssigned(wordsList) || wordsList.isEmpty();
    }
}

File: csharp/11CompleteImplementationCode/WordsFrequencyReporter.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace CompleteImplementationCode
{
	public class WordsFrequencyReporter
	{
		private const string EACH_LINE_REPORT_FORMAT = "{0} {1}\n";

		public StringBuilder Parse(int topNFrequency, String wordsList) {
			StringBuilder report = new StringBuilder();

			if (CheckIfWordsListIsNotAssignedOrEmpty(wordsList))
			{
				return report;
			}

			WordsFrequencyList wordsFrequencyHelper = new WordsFrequencyList();
			var wordsFrequencyList = wordsFrequencyHelper.CalculateWordsFrequency(wordsList)
				.OrderBy(word => word, new SortByFrequencyDescAndByWordAlphabetically())
				.ToList();

			return GenerateReportToReturn(report, wordsFrequencyList, topNFrequency);
		}

		private StringBuilder GenerateReportToReturn(StringBuilder report, ICollection<WordEntity> wordsFrequencyList, int topNFrequency)
		{
			if (topNFrequency <= 0)
			{
				return report;
			}

			topNFrequency = KeepTopNFrequencyWithinRange(wordsFrequencyList, topNFrequency);

			var filteredList = wordsFrequencyList.Take(topNFrequency);
			foreach (WordEntity wordEntity in filteredList)
			{
				report.AppendFormat(EACH_LINE_REPORT_FORMAT, wordEntity.Frequency, wordEntity.Word);
			}

			return report;
		}

		private int KeepTopNFrequencyWithinRange(ICollection<WordEntity> wordsFrequencyList, int topNFrequency)
		{
			return Math.Min(topNFrequency, wordsFrequencyList.Count);
		}

		private bool CheckIfWordsListIsNotAssignedOrEmpty(String wordsList) 
		{
			return WordsListIsNotAssigned(wordsList) || wordsList.Length == 0;
		}

		private bool WordsListIsNotAssigned(String wordsList) 
		{
			return wordsList == null;
		}
	}
}
